src/Shared/BrowserTesting/src/BrowserManagerConfiguration.cs

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.Configuration;
+using PlaywrightSharp;
+
+namespace Microsoft.AspNetCore.BrowserTesting
+{
+    public class BrowserManagerConfiguration
+    {
+        public BrowserManagerConfiguration(IConfiguration configuration)
+        {
+            Load(configuration);
+        }
+
+        public int TimeoutInMilliseconds { get; private set; }
+
+        public int TimeoutAfterFirstFailureInMilliseconds { get; private set; }
+
+        public string BaseArtifactsFolder { get; private set; }
+
+        public bool IsDisabled { get; private set; }
+
+        public LaunchOptions GlobalBrowserOptions { get; private set; }
+
+        public BrowserContextOptions GlobalContextOptions { get; private set; }
+
+        public IDictionary<string, BrowserOptions> BrowserOptions { get; } =
+            new Dictionary<string, BrowserOptions>(StringComparer.OrdinalIgnoreCase);
+
+        public ISet<string> DisabledBrowsers { get; } =
+            new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        public IDictionary<string, BrowserContextOptions> ContextOptions { get; private set; } =
+            new Dictionary<string, BrowserContextOptions>(StringComparer.OrdinalIgnoreCase);
+
+        public LaunchOptions GetLaunchOptions(LaunchOptions browserLaunchOptions)
+        {
+            if (browserLaunchOptions == null)
+            {
+                return GlobalBrowserOptions;
+            }
+            else
+            {
+                return Combine(GlobalBrowserOptions, browserLaunchOptions);
+            }
+        }
+
+        public BrowserContextOptions GetContextOptions(string browser)
+        {
+            if (!BrowserOptions.TryGetValue(browser, out var browserOptions))
+            {
+                throw new InvalidOperationException($"Browser '{browser}' is not configured.");
+            }
+            else if (browserOptions.DefaultContextOptions == null)
+            {
+                // Cheap clone
+                return Combine(GlobalContextOptions, null);
+            }
+            {
+                return Combine(GlobalContextOptions, browserOptions.DefaultContextOptions);
+            }
+        }
+
+        public BrowserContextOptions GetContextOptions(string browser, string contextName) =>
+            Combine(GetContextOptions(browser.ToString()), ContextOptions.TryGetValue(contextName, out var context) ? context : throw new InvalidOperationException("Invalid context name"));
+
+        public BrowserContextOptions GetContextOptions(string browser, string contextName, BrowserContextOptions options) =>
+            Combine(GetContextOptions(browser, contextName), options);
+
+        private void Load(IConfiguration configuration)
```

Comment:
You don't like using Options/ConfigBinding?

Reply:
I tried that, but it was painful, maybe I was missing something, can you define a "custom binder"?

The options types in playwright don't map cleanly to what I wanted (that's how I started) and I didn't want to create a completely separate schema.

Reply:
For example, I don't want to bind to arrays in the way config does, I need to convert those into objects and interpret the contents to make sure that we can override specific settings per environment if necessary.

Reply:
Also I haven't made this foolproof, I wrote it based on what I wanted us to be able to do through config and I stopped after I got things working.

I'm sure there are bugs and missing pieces here, but we don't need to do anything until we need them

Reply:
Yeah I guess if you are doing lots of nested complex objects you are left hand rolling, if we still owned the binder, seems worth considering making this a lot easier, because this is a ton of boilerplate.  There's the notion of associating typeconverters but unfortunately that's only for converting from a string.  At least the config apis are somewhat helpful.